File: 17 (Letter Combinations of a Phone Number)/python/Solution.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits: return []

        
        dic = {"2":"abc","3":'def',"4":"ghi","5":'jkl',"6":"mno","7":"pqrs","8":"tuv","9":"wxyz"}

        arr= []
        for ch in digits:
            arr.append(dic[ch])
        
        res = []
        def solve(idx,temp):
            if idx==len(arr):
                res.append(temp[::])
                return

            for i in range(len(arr[idx])):
                solve(idx+1,temp+arr[idx][i])
        
        solve(0,'')
        return res

File: 17 (Letter Combinations of a Phone Number)/cpp/Solution.cpp
class Solution {
public:

    void solve(int idx, string temp, vector<string> &res, vector<string> &arr){
        if (idx==arr.size()){
            res.push_back(temp);
            return;
        }

        for(int i = 0;i<arr[idx].size();i++){
            solve(idx+1,temp+arr[idx][i], res,arr);
        }
    }
    vector<string> letterCombinations(string digits) {
        if (digits.empty()) return {};

        unordered_map<char, string> dic =  {
            {'2', "abc"}, {'3', "def"}, {'4', "ghi"},
            {'5', "jkl"}, {'6', "mno"}, {'7', "pqrs"},
            {'8', "tuv"}, {'9', "wxyz"}
        };

        vector<string> arr;
        for(char ch: digits){
            arr.push_back(dic[ch]);
        }

        vector<string> res;
        solve(0,"",res,arr);
        return res;
    }
};